### DISCUSSION TITLE
Is binaryen the right tool for this?

### DISCUSSION THREAD

**Discussion by guest271314** [WebAssembly/binaryen#5749] [Q&A] [answered]
I am trying to compile eSpeak NG to WASM for use in the browser, to support SSML input and processing.

After building and isolating the files needed to successfully run `espeak-ng` where we pipe output to STDOUT with `--stdout` option the resulting folder and file structure is

- espeak-ng-data (folder)
- espeak-ng (executable)
- libespeak-ng.so.1.1.51 (shared library)
- libespeak-ng.so.1 (link to shared library)
- libespeak-ng.so (link to shared library)

I understand there is this work https://www.masswerk.at/mespeak/based off of https://github.com/kripken/speak.js.

I am trying to learn the _process_ of achieving the task.

How do I proceed cross-compiling to WASM for use in the browser?

**Comment by kripken** [comment, marked as answer]
Binaryen is a wasm->wasm optimizer. It sounds like you are porting a codebase to wasm, for which you need a toolchain like Emscripten. There is a tutorial which you might find useful:

https://emscripten.org/docs/getting_started/Tutorial.html

And here are docs for building something larger than a single file, which is I assume the case for espeak-ng:

https://emscripten.org/docs/compiling/Building-Projects.html

**Reply by guest271314** [reply]
The environment I am testing on is a temporary file system running on RAM. 

When I try to build Emscripten the system winds up running out of disk space. (Rust toolchain is over 1 GB, too, so when any crates, e.g., tokio, are downloaded I wind up running out of disk space). I have managed to download binaryen, and WASI-SDK where I did build a WASM file with C source code https://github.com/guest271314/native-messaging-webassembly.
 
I am trying to port eSpeak NG to WASM using the least amount of tools (resources) to do so, to determine the least amount of compilation resources to build a simple or complex universal executable targeting the browser.

I have isolated the minimal folders and files after compilation with `gcc` to execute the program. 

- espeak-ng-data (folder)
- espeak-ng (executable)
- libespeak-ng.so.1.1.51 (shared library)
- libespeak-ng.so.1 (link to shared library)

No audio is output, which is alright, I will just use `--stdout` to get output and parse the WAV in the browser both as a `MediaStreamTrack` and a downloadable file.

Is there a minimal version of Emscripten?

libespeak-ng.so (link to shared library)

**Reply by kripken** [reply]
> When I try to build Emscripten the system winds up running out of disk space.

Are you building it from source? If you can use prebuilt binaries that is much simpler, and takes much much less disk space,

https://emscripten.org/docs/getting_started/downloads.html

**Comment by guest271314** [comment]
Don't we still need to run 

```
# Fetch the latest version of the emsdk (not needed the first time you clone)
git pull

# Download and install the latest SDK tools.
./emsdk install latest

# Make the "latest" SDK "active" for the current user. (writes .emscripten file)
./emsdk activate latest

# Activate PATH and other environment variables in the current terminal
source ./emsdk_env.sh
```

which downloads among other programs Node.js `node-acorn`, etc?

**Reply by kripken** [reply]
Yes. But you mentioned "1 GB" before - I'm pretty sure all that together is much, much less than 1 GB.

**Reply by guest271314** [reply]
My experience is I run out of disk space. I'll try again.

**Reply by guest271314** [reply]
I started with 2.2 GB. After 

```
# Download and install the latest SDK tools.
./emsdk install latest

# Make the "latest" SDK "active" for the current user. (writes .emscripten file)
./emsdk activate latest
```

I have 1.1 GB of disk space.

I'll direct my inquiries over in Emscripten if I have issues compiling eSpeak NG to WASM. Thanks.

**Reply by guest271314** [reply]
I was able to build https://github.com/espeak-ng/espeak-ng/tree/master/emscripten with Emscripten using the information here https://github.com/espeak-ng/espeak-ng/issues/1543#issuecomment-1453844484, with 320 MB space on temporary file system to spare on the machine.


What I am trying to do is build eSpeak NG with with SSML support such as meSpeak.js https://www.masswerk.at/mespeak/ does which I am not sure how to do re updating the .cpp file, and IDL files. 

I also will update to use `MediaStreamTrackGenerator` and `AudioWorklet` versions instead of deprecated Web Audio API `ScriptProcessorNode` which that generated code does. I can handle that part on the front-end.

**Reply by guest271314** [reply]
These configurations wind up being very different from the original executable. What I would prefer is to expose the complete executable in the form of CLI usage.
